operators/src/upgrade/phases/init.rs

```diff
@@ -0,0 +1,249 @@
+use std::collections::HashMap;
+
+use crate::upgrade::k8s::crd::v0::UpgradePhase;
+
+/// Core components state.
+pub struct ComponentsState {
+    core: CoreComponentsState,
+
+    supportability: SupportabilityComponentsState,
+
+    tracing: TracingComponentsState,
+}
+
+impl ComponentsState {
+    /// Core components state.
+    pub(crate) fn core(&self) -> CoreComponentsState {
+        self.core.clone()
+    }
+
+    /// Supportability components state.
+    pub(crate) fn supportability(&self) -> SupportabilityComponentsState {
+        self.supportability.clone()
+    }
+
+    /// tracing components state.
+    pub(crate) fn tracing(&self) -> TracingComponentsState {
+        self.tracing.clone()
+    }
+
+    /// Initialize state.
+    pub(crate) fn with_state(upgrade_phase: UpgradePhase) -> Self {
+        Self {
+            core: CoreComponentsState::with_state(upgrade_phase.clone()),
+            supportability: SupportabilityComponentsState::with_state(upgrade_phase.clone()),
+            tracing: TracingComponentsState::with_state(upgrade_phase),
+        }
+    }
+
+    /// Covert into hashmaps.
+    pub(crate) fn convert_into_hash(&self) -> HashMap<String, HashMap<String, UpgradePhase>> {
+        let mut h: HashMap<String, HashMap<String, UpgradePhase>> = HashMap::new();
+        h.insert("core".to_string(), self.core().convert_into_hash());
+        h.insert(
+            "supportability".to_string(),
+            self.supportability().convert_into_hash(),
+        );
+        h.insert("tracing".to_string(), self.tracing().convert_into_hash());
+        h
+    }
+}
+
+/// Core components state.
+#[derive(Clone)]
+pub struct CoreComponentsState {
+    core_agent: UpgradePhase,
+    rest_api: UpgradePhase,
+    csi_controller: UpgradePhase,
+    csi_node: UpgradePhase,
+    io_agent: UpgradePhase,
+    metrics_exporter_pool: UpgradePhase,
+    disk_pool_operator: UpgradePhase,
+}
+
+impl CoreComponentsState {
+    /// Core agent state.
+    pub(crate) fn core_agent(&self) -> UpgradePhase {
+        self.core_agent.clone()
+    }
+
+    /// Rest api state.
+    pub(crate) fn rest_api(&self) -> UpgradePhase {
+        self.rest_api.clone()
+    }
+
+    /// Csi controller state.
+    pub(crate) fn csi_controller(&self) -> UpgradePhase {
+        self.csi_controller.clone()
+    }
+
+    /// Csi node state.
+    pub(crate) fn csi_node(&self) -> UpgradePhase {
+        self.csi_node.clone()
+    }
+
+    /// Io agent state.
+    pub(crate) fn io_agent(&self) -> UpgradePhase {
```

Comment:
There is no such thing as io_agent, do you mean io_engine?

Reply:
Renamed io_agent to io_engine.